pkg/middleware/rate_limit_test.go

```diff
@@ -0,0 +1,81 @@
+package middleware
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/macaron.v1"
+)
+
+type execFunc func() *httptest.ResponseRecorder
+type rateLimiterScenarioFunc func(c execFunc)
+
+func rateLimiterScenario(t *testing.T, desc string, rps int, burst int, fn rateLimiterScenarioFunc) {
+	t.Run(desc, func(t *testing.T) {
+		m := macaron.New()
+		defaultHandler := func(c *models.ReqContext) {
+			resp := make(map[string]interface{})
+			resp["message"] = "OK"
+			c.JSON(200, resp)
+		}
+		m.Use(macaron.Renderer(macaron.RenderOptions{
+			Directory: "",
+			Delims:    macaron.Delims{Left: "[[", Right: "]]"},
+		}))
+		m.Use(GetContextHandler(nil, nil, nil))
+		m.Get("/foo", RateLimit(rps, burst), defaultHandler)
+
+		fn(func() *httptest.ResponseRecorder {
+			resp := httptest.NewRecorder()
+			req, err := http.NewRequest("GET", "/foo", nil)
+			require.NoError(t, err)
+			m.ServeHTTP(resp, req)
+			return resp
+		})
+	})
+}
+
+func TestRateLimitMiddleware(t *testing.T) {
+	rateLimiterScenario(t, "rate limit calls, with burst", 10, 10, func(doReq execFunc) {
+		// first 10 calls succeed
+		for i := 0; i < 10; i++ {
+			resp := doReq()
+			assert.Equal(t, 200, resp.Code)
+		}
+
+		// next one fails
+		resp := doReq()
+		assert.Equal(t, 429, resp.Code)
+
+		// wait 1 for limiter tokens to appear, check that requests are accepted again
```

Comment:
```suggestion
		// wait 1 second for limiter tokens to appear, check that requests are accepted again
```

Reply:
Maybe we don't need to test that the rate library works as expected. Mocking rate limiter and  verify Middleware behaves as expected when Allow is called maybe enough? I think so

Reply:
I agree that we don't need to test external libs! However in this case I felt the need to test that my assumptions about how this lib behaves are correct. Call it an integration test  :)